saleor/attribute/utils.py

```diff
@@ -19,118 +22,169 @@
 T_INSTANCE = Union[Product, ProductVariant, Page]
 
 
-if TYPE_CHECKING:
-    from .models import AttributePage, AttributeProduct, AttributeVariant
+instance_to_function_variables_mapping = {
+    "Product": (
+        AttributeProduct,
+        AssignedProductAttribute,
+        AssignedProductAttributeValue,
+        "product",
+    ),
+    "ProductVariant": (
+        AttributeVariant,
+        AssignedVariantAttribute,
+        AssignedVariantAttributeValue,
+        "variant",
+    ),
+    "Page": (AttributePage, AssignedPageAttribute, AssignedPageAttributeValue, "page"),
+}
 
 
 def associate_attribute_values_to_instance(
-    instance: T_INSTANCE,
-    attribute: Attribute,
-    *values: AttributeValue,
-) -> AttributeAssignmentType:
-    """Assign given attribute values to a product or variant.
-
-    Note: be aware this function invokes the ``set`` method on the instance's
-    attribute association. Meaning any values already assigned or concurrently
+    instance: T_INSTANCE, attr_val_map: dict[int, list]
+):
+    """Assign given attribute values to a product, variant or page.
+
+    Note: be aware any values already assigned or concurrently
     assigned will be overridden by this call.
     """
-    values_ids = {value.pk for value in values}
 
     # Ensure the values are actually form the given attribute
-    validate_attribute_owns_values(attribute, values_ids)
+    validate_attribute_owns_values(attr_val_map)
 
     # Associate the attribute and the passed values
-    assignment = _associate_attribute_to_instance(instance, attribute.pk)
-    assignment.values.set(values)
-
-    # While migrating to a new structure we need to make sure we also
-    # copy the assigned product to AssignedProductAttributeValue
-    # where it will live after issue #12881 will be implemented
-    if isinstance(instance, Product):
-        AssignedProductAttributeValue.objects.filter(
-            assignment_id=assignment.pk
-        ).update(product_id=instance.pk)
-
-    # This code will be deleted in new release (3.17), it is temporary solution between
-    # releases to keep database in sync
-    elif isinstance(instance, Page):
-        AssignedPageAttributeValue.objects.filter(assignment_id=assignment.pk).update(
-            page_id=instance.pk
-        )
+    _associate_attribute_to_instance(instance, attr_val_map)
 
-    sort_assigned_attribute_values(instance, assignment, values)
 
-    return assignment
+def _associate_attribute_to_instance(
+    instance: T_INSTANCE, attr_val_map: dict[int, list]
+):
+    instance_type = instance.__class__.__name__
+    variables = instance_to_function_variables_mapping.get(instance_type)
+
+    if not variables:
+        raise AssertionError(f"{instance_type} is unsupported")
```

Comment:
maybe TypeError would be more accurate? If it's a big change we can skip it for now.